File: libs/expo-pdf-high-res-generator/android/src/main/java/expo/modules/pdfhighresgenerator/ExpoPdfHighResGeneratorModule.kt
package expo.modules.pdfhighresgenerator

import android.graphics.Bitmap
import android.graphics.pdf.PdfRenderer
import androidx.core.net.toUri
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

private class PdfGenerationException(message: String, cause: Throwable? = null) :
  CodedException(message, cause)

class ExpoPdfHighResGeneratorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoPdfHighResGenerator")

    AsyncFunction("generate") { filePath: String, scale: Double ->
      generate(filePath, scale)
    }
  }

  // Renders every page of the PDF at `filePath` to a JPEG, returning the `file://` URIs of the produced images.
  private fun generate(filePath: String, scale: Double): List<String> {
    var fileDescriptor: android.os.ParcelFileDescriptor? = null
    var tempPdfFile: File? = null
    var bitmap: Bitmap? = null
    try {
      val context = appContext.reactContext
        ?: throw PdfGenerationException("React context is not available")
      val uri = filePath.toUri()
      // Needed to resolve permission issues and "content://" vs "file://"
      tempPdfFile = File(context.cacheDir, "temp_source_${UUID.randomUUID()}.pdf")
      val inputStream = context.contentResolver.openInputStream(uri)
        ?: throw PdfGenerationException("Unable to open stream for: $filePath")

      // Copy stream -> file
      FileOutputStream(tempPdfFile).use { outputStream ->
        inputStream.use { input ->
          input.copyTo(outputStream)
        }
      }

      // Use it now that we have a proper file
      fileDescriptor = android.os.ParcelFileDescriptor.open(
        tempPdfFile,
        android.os.ParcelFileDescriptor.MODE_READ_ONLY
      )
      val renderer = PdfRenderer(fileDescriptor)
      val outputPaths = mutableListOf<String>()
      val cacheDir = context.cacheDir

      // Maximum dimensions to prevent memory issues
      val maxWidth = 4096
      val maxHeight = 4096

      for (i in 0 until renderer.pageCount) {
        val page = renderer.openPage(i)

        // Calculate target size with initial scale
        var width = (page.width * scale).toInt()
        var height = (page.height * scale).toInt()

        // Apply limits: scale down if exceeds max dimensions
        if (width > maxWidth || height > maxHeight) {
          val widthScale = maxWidth.toDouble() / width
          val heightScale = maxHeight.toDouble() / height
          val limitScale = minOf(widthScale, heightScale)
          width = (width * limitScale).toInt()
          height = (height * limitScale).toInt()
        }

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        // Set white background for better visibility
        bitmap.eraseColor(android.graphics.Color.WHITE)

        page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)

        val fileName = "pdf_render_${UUID.randomUUID()}_$i.jpg"
        val outputFile = File(cacheDir, fileName)
        val outputStream = FileOutputStream(outputFile)

        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, outputStream)
        outputStream.close()

        outputPaths.add("file://${outputFile.absolutePath}")

        page.close()
        bitmap.recycle()
      }

      renderer.close()
      return outputPaths
    } catch (e: OutOfMemoryError) {
      throw PdfGenerationException("Out of memory while rendering PDF", e)
    } catch (e: PdfGenerationException) {
      throw e
    } catch (e: Exception) {
      throw PdfGenerationException(e.message ?: "Unknown error while rendering PDF", e)
    } finally {
      bitmap?.recycle()
      try {
        fileDescriptor?.close()
        tempPdfFile?.delete()
      } catch (e: Exception) {
        // Ignore errors on close
      }
    }
  }
}

File: libs/expo-pdf-high-res-generator/ios/ExpoPdfHighResGeneratorModule.swift
import ExpoModulesCore
import CoreGraphics
import UIKit

public class ExpoPdfHighResGeneratorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoPdfHighResGenerator")

    AsyncFunction("generate") { (filePath: String, scale: Double) -> [String] in
      try Self.generate(filePath: filePath, scale: scale)
    }
    .runOnQueue(DispatchQueue.global(qos: .userInitiated))
  }

  // Renders every page of the PDF at `filePath` to a JPEG, returning the `file://` URIs of the produced images.
  private static func generate(filePath: String, scale: Double) throws -> [String] {
    // Copy source to a temporary file to ensure safe/fast access without sandbox locks.
    let fileManager = FileManager.default
    let tempDir = fileManager.temporaryDirectory
    let tempSourceId = UUID().uuidString
    let tempSourceUrl = tempDir.appendingPathComponent("temp_source_\(tempSourceId).pdf")

    let sourceUrl: URL? = filePath.hasPrefix("file://")
      ? URL(string: filePath)
      : URL(fileURLWithPath: filePath)

    guard let validSourceUrl = sourceUrl else {
      throw PdfGenerationException("Invalid path: \(filePath)")
    }

    do {
      if fileManager.fileExists(atPath: tempSourceUrl.path) {
        try fileManager.removeItem(at: tempSourceUrl)
      }
      try fileManager.copyItem(at: validSourceUrl, to: tempSourceUrl)
    } catch {
      throw PdfGenerationException("Could not copy to temp: \(error.localizedDescription)")
    }

    // We use CGPDFDocument instead of PDFKit for maximum performance and no UI overhead.
    guard let pdfDoc = CGPDFDocument(tempSourceUrl as CFURL) else {
      try? fileManager.removeItem(at: tempSourceUrl)
      throw PdfGenerationException("Could not open PDF with CoreGraphics")
    }

    var outputPaths: [String] = []
    let pageCount = pdfDoc.numberOfPages // CGPDF pages are 1-based

    // Maximum dimensions to prevent memory issues
    let maxWidth = 4096.0
    let maxHeight = 4096.0

    for i in 1...pageCount {
      autoreleasepool {
        guard let page = pdfDoc.page(at: i) else { return }

        // Get Box Rect (MediaBox is the physical page size)
        let pageRect = page.getBoxRect(.mediaBox)

        // Calculate Target Size with initial scale
        var targetWidth = pageRect.width * CGFloat(scale)
        var targetHeight = pageRect.height * CGFloat(scale)

        // Apply limits: scale down if exceeds max dimensions
        var finalScale = CGFloat(scale)
        if targetWidth > maxWidth || targetHeight > maxHeight {
          let widthScale = maxWidth / targetWidth
          let heightScale = maxHeight / targetHeight
          let limitScale = min(widthScale, heightScale)
          finalScale *= limitScale
          targetWidth *= limitScale
          targetHeight *= limitScale
        }

        let finalWidth = Int(targetWidth)
        let finalHeight = Int(targetHeight)

        // Create Bitmap Context, 8 bits per component, 4 components (ARGB), standard RGB color space
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bitmapInfo = CGImageAlphaInfo.premultipliedFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue

        guard let context = CGContext(data: nil,
                                      width: finalWidth,
                                      height: finalHeight,
                                      bitsPerComponent: 8,
                                      bytesPerRow: 0,
                                      space: colorSpace,
                                      bitmapInfo: bitmapInfo) else { return }
        // Fill white background
        context.setFillColor(UIColor.white.cgColor)
        context.fill(CGRect(x: 0, y: 0, width: CGFloat(finalWidth), height: CGFloat(finalHeight)))

        context.interpolationQuality = .high
        context.saveGState()

        // Scale context to match the target resolution
        context.scaleBy(x: finalScale, y: finalScale)

        // We need to translate so the mediaBox origin (which might be non-zero) aligns with 0,0
        context.translateBy(x: -pageRect.origin.x, y: -pageRect.origin.y)

        context.drawPDFPage(page)
        context.restoreGState()

        // --- SAVE IMAGE ---
        if let cgImage = context.makeImage() {
          let uiImage = UIImage(cgImage: cgImage)
          if let data = uiImage.jpegData(compressionQuality: 0.9) {
            let fileName = "pdf_render_\(UUID().uuidString)_\(i).jpg"
            let fileURL = tempDir.appendingPathComponent(fileName)

            do {
              try data.write(to: fileURL)
              outputPaths.append(fileURL.absoluteString)
            } catch {
              print("Error saving page \(i): \(error)")
            }
          }
        }
      }
    }

    // CLEANUP
    try? fileManager.removeItem(at: tempSourceUrl)

    return outputPaths
  }
}

private final class PdfGenerationException: Exception {
  private let message: String

  init(_ message: String) {
    self.message = message
    super.init()
  }

  override var reason: String {
    message
  }
}
